File: 4 Kyu/4 kyu -Square into Squares. Protect trees!.cpp
#include <vector>
#include <algorithm>
class Decomp {
  public:
      static std::vector<long long> decompose(long long n) {

          long long n2 = 0, num;
          std::vector<long long> V = {n};

          do {
              num = V.back ();
              V.pop_back();
              n2 += num * num;

              while (num-->1)
                  if (num * num <= n2) {
                      V.push_back (num);
                      n2 -= num * num;
                  }

          } while (n2 != 0 && V.size() != 0);
          sort (V.begin(), V.end());
          return V;
     }
};

File: 4 Kyu/4 kyu - Square into Squares. Protect trees!.rs
fn decompose(n: i64) -> Option<Vec<i64>> {

    let mut n2:i64 = 0;
    let mut vs = vec![n];

    loop  {
        let mut num = *vs.last().unwrap();
        vs.pop();
        n2 += num * num;

        while num > 1 {
            num -= 1;
            if num * num <= n2 {
                vs.push (num);
                n2 -= num * num;
            }
        }

        if n2 == 0 || vs.len() == 0 { break }
    }

    vs.sort();

    if vs.len() == 0 { return None }
    Some (vs)
}
